// maxim-workflow/index.js
function Qcloud(){}
Qcloud.prototype = {
    copyFiles : function(files,config,callback){
        //只做文件在本地目录上的迁移和拷贝，这里是将指定源目录下的文件拷贝到指定的目标目录下
        var mover = require('./lib/mover.js');
        //拷贝文件
        new mover().copy(files,config,callback);
    },
    sprite : function (imgs,config,callback) {
        //对指定的样式文件进行雪碧图合并
        var spriter = require('./lib/sprite.js');
        new spriter().sprite(imgs,config,callback);
    },
    miniCsses : function(csses,config,callback){
        //调用lib的自定义库
        var min = require('./lib/mincss.js');
        new min().minCsses(csses,config,callback);
    },
    ftpUtil : function(files,Config,callback){
        var ftpUtil = require('./lib/ftpUtil.js');
        new ftpUtil().ftp(files,Config,callback);
    },
    tinyImg : function(imgs,Config,globalConfig,callback){
        var tinyImg = require('./lib/tinyimg.js');
        new tinyImg().TinyImg(imgs,Config,globalConfig,callback);
    },
    youtu : function(imgs,Config,globalConfig,callback){
        var youtu = require('./lib/youtuCompressor.js');
        new youtu().compressor(imgs,Config,globalConfig,callback);
    }
};


module.exports = Qcloud;
