// maxim-workflow/lib/mover.js
function Copyfile(){}
Copyfile.prototype = {
    /**
     * flag ：为空时表示files都是在源目录下的文件
     * */
    copy : function(files,config,callback){
        if(files.length <= 0){
            return;
        }
        //预处理，如果传入的是单个文件名将其转换成文件字符串
        if(typeof files === 'string'){
            var tmp = files;
            files = [];
            files.push(tmp);
        }
        //变量定义
        var util = require('./common.js');
        var plat = util.getCurrentPlat();
        var split = plat.split;

        var fs = require('fs');
        var path = require('path');
        var results = [];
        var count = files.length;
        var index = count;
        //时间戳：要求拷贝的资源打版本且打版本的开关是开的
        var timestampSprite = config.timeStampSwitch && config.imgSyncSwitch === 'true' ? '-'+config.imgSyncName : '';
        //处理config中路径的配置如果不以\结尾，则要加上
        config.localPath = config.localPath[config.localPath.length-1] === split ? config.localPath : config.localPath +split;
        config.destPath = config.destPath[config.destPath.length-1] === split ? config.destPath : config.destPath + split;

        //循环拷贝文件
        for(var i = 0;i<count;i++){
            copyfile(fs,i)();
        }
        function copyfile(fs,k){
            return function () {
                //创建目标路径，判断每一个文件的路径
                var realDest = files[k].replace(config.localPath,config.destPath).replace(path.basename(files[k]),'');
                util.createPath(realDest,fs);

                var rname = files[k].replace(config.localPath,config.destPath);
                //判断是否添加时间戳
                var rname = path.dirname(rname) + path.sep + path.basename(rname).replace(path.extname(rname),'') + timestampSprite + path.extname(rname);
                try{
                    if(!fs.existsSync(files[k])){
                        results.push({
                            fName : rname.replace(config.destPath,'').replace(/\\/g,'\/'),
                            status : false,
                            message : 'file is not existed'
                        });
                    }else{
                        if(files[k] !== rname){
                            //源文件和目标文件不是同一个路径时，执行流处理
                            fs.createReadStream(files[k]).pipe(fs.createWriteStream(rname));
                        }
                        results.push({
                            fName : rname.replace(config.destPath,'').replace(/\\/g,'\/'),
                            status : true
                        });
                    }
                }catch(e){
                    results.push({
                        fName : rname.replace(config.destPath,'').replace(/\\/g,'\/'),
                        status : false,
                        message : e.message
                    });
                }finally{
                    index--;
                    if(index <= 0){
                        callback(results);
                    }
                }
            }
        }
    },
    move : function(files,destPath,callback){

    }
};
module.exports = Copyfile;
