// maxim-workflow/lib/svn.js
/**
 * Created by owenhong on 2016/8/8.
 */

'use strict';

var os = require("os");
var path = require("path");
var Client = require('svn-spawn');
var fse = require('fs-extra');




function svn(){}

svn.prototype.init = function(files,Config,callback){
    var client = new Client({
        cwd: Config.svnLocalPath,
        username: Config.svnUser, // optional if authentication not required or is already saved
        password: Config.svnPassword, // optional if authentication not required or is already saved
        noAuthCache: false, // optional, if true, username does not become the logged in user on the machine
    });

    client.update(function(err, data) {
        if (err) {
            console.log('update SVN...');
            console.log(err);
        }
        console.log('updated success...');

        client.addLocal(function (err, data) {
            if (err) {
                console.log('addLocal error....')
                console.log(err);
            }

            console.log('all local changes has been added for commit');


            var results = [];
            ////判断文件是否存在，不存在创建文件夹
            //if(os.type() == "Windows_NT"){
            //    var filePath = result.path.replace(/\//g, '\\');
            //}else{
            //    var filePath = result.path;
            //}
            //var $fName = filePath.replace(config.destPath, '').replace(config.localPath, '').replace(/\\/g, '\/');

            client.commit('Maxim commit', function(err, data) {
                if (err) {
                    console.log('commit error....')
                    if(err.message.indexOf('E155015') >= 0) {

                        console.log('E155015...');

                        let length = files.length;
                        let index = length;

                        files.forEach(function (filePath) {
                            fse.remove(filePath, function (err) {
                                if (err) {
                                    index--;
                                    if (index <= 0) {
                                        results.push({
                                            svnSwitch: true,
                                            status: false,
                                            message: "SVN文件提交时出现冲突"
                                        });
                                        callback(results);
                                    }
                                }

                                index--;
                                if (index <= 0) {
                                    client.update(function(err, data) {
                                        if (err) {
                                            console.log('update SVN...');
                                            console.log(err);

                                            results.push({
                                                svnSwitch: true,
                                                status: false,
                                                message: "SVN文件提交时出现冲突"
                                            });
                                            callback(results);
                                        }
                                        results.push({
                                            svnSwitch: true,
                                            status: false,
                                            message: "SVN文件提交时出现冲突"
                                        });
                                        callback(results);
                                    });
                                }
                            });
                        });
                    }else{
                        console.log('No E155015...');
                        results.push({
                            svnSwitch: true,
                            fName: data,
                            status: false,
                            message: err.message
                        });

                        callback(results);
                    }
                }else{
                    console.log('goods.....');
                    results.push({
                        svnSwitch: true,
                        status: true,
                        message: 'svn commit success...'
                    });

                    callback(results);
                }
            });
        });
    });
}


module.exports = svn;
